lib/client-api/src/hooks.ts

```diff
@@ -251,3 +266,45 @@ export function useEffect(create: () => (() => void) | void, deps?: any[]): void
   const effect = useMemoLike('useEffect', () => ({ create }), deps);
   currentEffects.push(effect);
 }
+
+export interface Listener {
+  (...args: any[]): void;
+  ignorePeer?: boolean;
+}
+
+export interface EventMap {
+  [eventId: string]: Listener;
+}
+
+export function useChannel(eventMap: EventMap) {
+  const channel = addons.getChannel();
+  useEffect(() => {
+    Object.entries(eventMap).forEach(([type, listener]) => channel.on(type, listener));
+    return () => {
+      Object.entries(eventMap).forEach(([type, listener]) =>
+        channel.removeListener(type, listener)
+      );
+    };
+  }, Object.keys(eventMap));
```

Comment:
So if I change the listener function, it has no effect?

Reply:
@Hypnosphi yeah I'm a little unsure.. I think this api would be nice:

```js
import { useChannel } from '@storybook/client-api';

export const myDecorator = () => {

  const emit = useChannel({
    [EVENTS.REQUEST]: () => {
      emit(EVENTS.RESULT, ['from the preview'])}
  });
};
```

but with the function definition being like that, that would mean, the function would get recreated every render, and thus the `useEffect` would remove and re-add on every render, wouldn't it? That's what I'm trying to avoid.

My assumption would be that users would map an eventName to a handlerFunction, and that mapping wouldn't change after defining it once. Do you think my assumption is wrong?

I think this would be the only alternative then:

```js
import { useChannel } from '@storybook/client-api';

const map = {
  [EVENTS.REQUEST]: () => {
    emit(EVENTS.RESULT, ['from the preview'])}
  },
};

export const myDecorator = () => {
  const emit = useChannel(map);
};
```

This way the reference to the entire object is always the same.

It makes it quite a bit more involved to use though. WDYT?

Reply:
export function useChannel(eventMap: EventMap, deps?: any[]) {
  ...
  useEffect(() => {
    ...
  }, [...Object.keys(eventMap), ...deps]);

Reply:
> and thus the useEffect would remove and re-add on every render, wouldn't it?

@ndelangen that's what we do in manager anyway, so maybe it's not that bad: https://github.com/storybookjs/storybook/blob/next/lib/api/src/index.tsx#L341-L346